sympy/polys/matrices/domainmatrix.py

```diff
@@ -148,6 +163,24 @@ def mul(A, b):
     def matmul(A, B):
         return A.from_rep(A.rep.matmul(B.rep))
 
+    def scalarmul(A, lamda):
+        if not isinstance(lamda, DomainScalar):
+            lamda = DomainScalar.from_sympy(lamda)
```

Comment:
There should be no automatic conversion from Expr. We shouldn't let DomainMatrix interact with Expr directly without explicit conversions.

Reply:
I think DomainMatrix should only be able to interact with DomainScalar or `int`.

Reply:
> There should be no automatic conversion from Expr. We shouldn't let DomainMatrix interact with Expr directly without explicit conversions.
>
Maybe I didn't understand quite well.
 But here, `lamda` is an `Expr`, and not `DomainMatrix`, thus using `construct_domain`, we want to convert the `Expr` to `DomainScalar`. So there will no interaction between DomainMatrix and Expr, which is one of the keys for introducing `DomanScalar`.

Reply:
Sorry, I got the point. 
The issue is how to check for the `lamda` such that it should be either `Sympy Expr` or `Domain element`, not something like `DDM` or `DomainMatrix`.

Not sure if this should be the way:
```
if not isinstance(lamda, DomainScalar):
    if not isinstance(lamda, Expr):
        if isinstance(lamda, DomainMatrix) or isinstance(lamda, DDM) or isinstance(lamda, SDM):
            return TypeError
    lamda = DomainScalar.from_sympy(lamda)
```
@oscarbenjamin, Kindly share your opinion.
Thanks.

Reply:
Expr should not be allowed here. I would only allow a plain Python `int` or a `DomainScalar`.

Reply:
Actually only `DomainScalar` should be allowed in this method but `__mul__` could also allow an `int` (and convert it to `DomainScalar` before calling this)

Reply:
> Expr should not be allowed here. I would only allow a plain Python `int` or a `DomainScalar`.

Sure, will remove it.

> Actually only `DomainScalar` should be allowed in this method but `__mul__` could also allow an `int` (and convert it to `DomainScalar` before calling this)


But what if we pass someone passes a `DomainMatrix` or `DDM` or `SDM` perhaps. We will convert the `Expr` or any `Domain element` to `DomainScalar` before using `SDM.scalarmul`. 
Moreover, In the latest commit, the `SDM.scalarmul` uses `unop_dict`, while `SDM.scalarmul` uses `mul`.

Kindly share your opinion.
Thanks.

Reply:
> But what if we pass someone passes a `DomainMatrix` or `DDM` or `SDM` perhaps.

Then they should get a `TypeError`. This is `scalarmul` so only scalars should be allowed.

There should also be a `TypeError` when passing `Expr` or a domain element as well. It's much better to have a `TypeError` so you can correct your code to convert between `Expr`, domain element, `DomainScalar` etc in the right place.

The only exception that I think can be reasonable for convenience is allowing interoperation with `int` e.g. `2*M`. Converting that `int` to `DomainScalar` can be done in `__mul__` though. The point of methods like `scalarmul` is that they are supposed to be able to assume that any conversions have already happened so we can also use them if we want to skip the cost of checking/conversion.

This method should just assume that the input is a `DomainScalar`. The `__mul__` method can check for `DomainScalar` before calling this.

Reply:
> There should also be a `TypeError` when passing `Expr` or a domain element as well. It's much better to have a `TypeError` so you can correct your code to convert between `Expr`, domain element, `DomainScalar` etc in the right place.
> 
I am not sure if there is a need for `TypeError` for `Expr `as it can be converted to `DomainScalar` easily, further `scalarmul`
 can be called.

> The only exception that I think can be reasonable for convenience is allowing interoperation with `int` e.g. `2*M`. Converting that `int` to `DomainScalar` can be done in `__mul__` though. 
> 
The case of `int * DomainMatrix` already handled by `SDM.__mul__`. 
The problem is if we perform the operation with something other than `int`, it raises `TypeError`
```
In [2]: dm=DomainMatrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]], (3, 3), ZZ)

In [3]: 1.5*dm
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-ab95fb13257e> in <module>
----> 1 1.5*dm

TypeError: unsupported operand type(s) for *: 'float' and 'DomainMatrix'
```
I am not sure if we want to handle it. If yes, then how should we handle it since the traceback doesn't help us to find where the issue is raised.

Kindly provide your opinion,
Thanks.